mdb/debug_client.py

```diff
@@ -38,33 +42,43 @@ async def init_debug_proc(self):
         for command in backend.start_commands:
             dbg_proc.sendline(command)
             await dbg_proc.expect(backend.prompt_string, async_=True)
-        print(f"{backend.name} initialized")
+
+        logger.debug("Backend init finished: %s", backend.name)
         self.dbg_proc = dbg_proc
 
     def interact(self, message):
         if message["rank"] == self.myrank:
-            logging.info(f"running interact mode on rank {self.myrank}.")
+            logger.info("Running interact mode on rank %s", self.myrank)
             self.dbg_proc.interact()
 
-    async def wait_for_command(self):
-        message = await self.conn.recv_message()
-        command = message["command"]
+    async def execute_command(self, command, prev: asyncio.Task):
         if command == "interrupt":
-            print("HEEEEEEEEEEYYYYYYYYYYYYYYYYYY!!!!!!!!!!!")
+            logger.warning("Interrupt received")
+            # stop whatever is current running, so it doesn't try to reply
+            success = prev.cancel()
+            # send intterupt to the process
+            self.dbg_proc.sendintr()
+            await self.dbg_proc.expect(self.backend.prompt_string, async_=True)
+            # report on how that all went
+            reply = {
+                "result": f"Interrupted: {success}",
+                "rank": self.myrank,
+            }
+
         else:
-            print("running command :: ", command)
+            logger.debug("Running command: '%s'", command)
             self.dbg_proc.sendline(command)
             await self.dbg_proc.expect(self.backend.prompt_string, async_=True)
 
             result = self.dbg_proc.before.decode()
             result = strip_bracketted_paste(result)
-
-            message = {
+            reply = {
                 "result": result,
                 "rank": self.myrank,
             }
 
-            await self.conn.send_message(message)
+        logger.debug("Replying to: '%s'", command)
+        await self.conn.send_message(reply)
```

Comment:
I did some testing with the `simple-mpi.f90` program and realised this needs to change a bit.

The problem is for the processes that don't "need" interrupting (because they have already finished running) the output gets out-of-sync. So we only want to send a message when the interrupt is required. Because an async message will have already been received for the finished tasks. Also we need to clear the buffer (using `before`) so that when we run the next valid command the output doesn't contain the `SIGINT` stuff.

I have an idea how to do it, so I might just accept your changes and then make another PR rather than requesting changes.

Essentially it is something like:

```suggestion
        if command == "interrupt":
            logger.warning("Interrupt received")
            # stop whatever is current running, so it doesn't try to reply
            success = prev.cancel()
            logger.info(f"success = {success}")
            # send intterupt to the process
            if success:
                self.dbg_proc.sendintr()
                self.dbg_proc.expect(self.backend.prompt_string)
                _ = self.dbg_proc.before.decode()
            # report on how that all went
            reply = {
                "result": f"Interrupted: {success}",
                "rank": self.myrank,
            }

            logger.debug("Replying to: '%s'", command)
            if success:
                await self.conn.send_message(reply)
```

Reply:
Ah good point! This was another oversight of mine.

This should work fine but I slightly refactored how the `reply` is sent, so I'll add a commit that implements your changes in a moment!